awx/main/scheduler/task_manager.py

```diff
@@ -267,184 +262,17 @@ class DependencyManager(TaskBase):
-            if "inventory_sources_already_updated" in start_args and inventory_source.id in start_args['inventory_sources_already_updated']:
-                continue
-            if not inventory_source.update_on_launch:
-                continue
-            latest_inventory_update = self.get_latest_inventory_update(inventory_source)
-            if self.should_update_inventory_source(task, latest_inventory_update):
-                inventory_task = self.create_inventory_update(task, inventory_source)
-                created_dependencies.append(inventory_task)
-                dependencies.append(inventory_task)
-            else:
-                dependencies.append(latest_inventory_update)
-
-        if dependencies:
-            self.add_dependencies(task, dependencies)
-
-        return created_dependencies
-
-    def gen_dep_for_inventory_update(self, inventory_task):
-        created_dependencies = []
-        if inventory_task.source == "scm":
-            invsrc = inventory_task.inventory_source
-            if not invsrc.source_project.scm_update_on_launch:
-                return created_dependencies
-
-            latest_src_project_update = self.get_latest_project_update(invsrc.source_project_id)
-            if self.should_update_related_project(inventory_task, latest_src_project_update):
-                latest_src_project_update = self.create_project_update(inventory_task, project_id=invsrc.source_project_id)
-                created_dependencies.append(latest_src_project_update)
-            self.add_dependencies(inventory_task, [latest_src_project_update])
-            latest_src_project_update.scm_inventory_updates.add(inventory_task)
-        return created_dependencies
-
-    @timeit
-    def generate_dependencies(self, undeped_tasks):
-        created_dependencies = []
-        for task in undeped_tasks:
-            task.log_lifecycle("acknowledged")
-            if type(task) is Job:
-                created_dependencies += self.gen_dep_for_job(task)
-            elif type(task) is InventoryUpdate:
-                created_dependencies += self.gen_dep_for_inventory_update(task)
-            else:
-                continue
-        UnifiedJob.objects.filter(pk__in=[task.pk for task in undeped_tasks]).update(dependencies_processed=True)
-
-        return created_dependencies
-
-    def process_tasks(self):
-        deps = self.generate_dependencies(self.all_tasks)
-        self.generate_dependencies(deps)
-        self.subsystem_metrics.inc(f"{self.prefix}_pending_processed", len(self.all_tasks) + len(deps))
-
     @timeit
     def _schedule(self):
-        self.get_tasks(dict(status__in=["pending"], dependencies_processed=False))
-
-        if len(self.all_tasks) > 0:
-            self.get_inventory_source_tasks()
-            self.process_tasks()
+        r = (
+            UnifiedJob.objects.filter(status='pending', dependencies_processed=False)
+            .exclude(dependent_jobs__status__in=ACTIVE_STATES)
```

Comment:
I realized that this might not actually be correct. Instead of `ACTIVE_STATES`, we only want `"successful"`, because canceled / error / failed dependencies should not progress a job to the next phase in the task manager kind of methods.

@fosterseth thinking back to https://github.com/ansible/awx/issues/12999, it wouldn't be hard to have the dependency manager absorb the work that the `handle_work_error` type of errback methods do now. We could do this update here for successful dependencies, and then do a second query where we get a list of all jobs that are both (1) in pending status and (2) have dependencies with failure status. Since you already have this running on a schedule, it would be a fool-proof way to mop up any cascading failures. Just an idea.

Reply:
I implemented what I mentioned here, but haven't removed the errbacks... following this through...

We pass dependent_jobs in `start_task`, but those basically come from `get_jobs_fail_chain` in the models.

In `TaskManagerInventoryUpdateMixin` and `TaskManagerProjectUpdateMixin` these pretty much equate to returning the reverse relationship objects - `unifiedjob_blocked_jobs`. The only exception is jobs giving related `project_update`, in which case I'm _sure_ we don't rely on this.

Having the `DependencyManager` run on a schedule is as airtight as we can get.

So to summarize, I'm ready to delete the use of `errbacks`, `handle_work_error`, and the vast network of inheritance starting from `TaskManagerUnifiedJobMixin`. This is a lot to nuke, so I'd like to check with others before fully carrying that out.

Reply:
green light from me